exp/controllers/gcpmanagedcontrolplane_controller.go

```diff
@@ -29,33 +50,142 @@ import (
+		log.Info("Reconciliation is paused for this object")
+		return ctrl.Result{}, nil
+	}
+
+	managedControlPlaneScope, err := scope.NewManagedControlPlaneScope(scope.ManagedControlPlaneScopeParams{
+		Client:                 r.Client,
+		Cluster:                cluster,
+		GCPManagedControlPlane: gcpManagedControlPlane,
+	})
+	if err != nil {
+		return ctrl.Result{}, errors.Errorf("failed to create scope: %+v", err)
+	}
 
-	// TODO(user): your logic here
+	// Always close the scope when exiting this function so we can persist any GCPMachine changes.
+	defer func() {
+		if err := managedControlPlaneScope.Close(); err != nil && reterr == nil {
+			reterr = err
+		}
+	}()
+
+	// Handle deleted clusters
+	if !gcpManagedControlPlane.DeletionTimestamp.IsZero() {
+		return r.reconcileDelete(ctx, managedControlPlaneScope)
+	}
+
+	// Handle non-deleted clusters
+	return r.reconcile(ctx, managedControlPlaneScope)
+}
+
+func (r *GCPManagedControlPlaneReconciler) reconcile(ctx context.Context, managedControlPlaneScope *scope.ManagedControlPlaneScope) (ctrl.Result, error) {
+	log := log.FromContext(ctx)
+	log.Info("Reconciling GCPManagedControlPlane")
+
+	controllerutil.AddFinalizer(managedControlPlaneScope.GCPManagedControlPlane, infrav1.ClusterFinalizer)
+	if err := managedControlPlaneScope.PatchObject(); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	reconcilers := []cloud.Reconciler{
+		clusters.New(managedControlPlaneScope),
+	}
+
+	for _, r := range reconcilers {
+		if err := r.Reconcile(ctx); err != nil {
+			log.Error(err, "Reconcile error")
+			record.Warnf(managedControlPlaneScope.GCPManagedControlPlane, "GCPManagedControlPlaneReconcile", "Reconcile error - %v", err)
+			return ctrl.Result{}, err
+		}
+	}
+
+	if conditions.IsTrue(managedControlPlaneScope.GCPManagedControlPlane, infrav1exp.GKEControlPlaneCreatingCondition) {
+		return ctrl.Result{RequeueAfter: 5 * time.Minute}, nil
+	}
 
 	return ctrl.Result{}, nil
 }
 
-// SetupWithManager sets up the controller with the Manager.
-func (r *GCPManagedControlPlaneReconciler) SetupWithManager(mgr ctrl.Manager) error {
-	return ctrl.NewControllerManagedBy(mgr).
-		For(&infrastructurev1beta1.GCPManagedControlPlane{}).
-		Complete(r)
+func (r *GCPManagedControlPlaneReconciler) reconcileDelete(ctx context.Context, managedControlPlaneScope *scope.ManagedControlPlaneScope) (ctrl.Result, error) {
+	log := log.FromContext(ctx)
+	log.Info("Deleting GCPManagedControlPlane")
+
+	reconcilers := []cloud.Reconciler{
+		clusters.New(managedControlPlaneScope),
+	}
+
+	for _, r := range reconcilers {
+		if err := r.Delete(ctx); err != nil {
+			log.Error(err, "Reconcile error")
+			record.Warnf(managedControlPlaneScope.GCPManagedControlPlane, "GCPManagedControlPlaneReconcile", "Reconcile error - %v", err)
+			return ctrl.Result{}, err
+		}
+	}
+
+	if conditions.Get(managedControlPlaneScope.GCPManagedControlPlane, infrav1exp.GKEControlPlaneDeletingCondition).Reason == infrav1exp.GKEControlPlaneDeletedReason {
```

Comment:
This may be better handled by the return from the reconcilers

Reply:
done